multigeiger/tube.cpp

```diff
@@ -35,22 +35,121 @@ TUBETYPE tubes[] = {
 
 volatile bool isr_GMC_cap_full;
 
+volatile unsigned long isr_hv_pulses;
+volatile bool isr_hv_charge_error;
+
 volatile unsigned int isr_GMC_counts;
 volatile unsigned long isr_count_timestamp;
 volatile unsigned long isr_count_time_between;
 
-static unsigned long hvpulse_timestamp;  // initialized via setup_tube -> gen_charge_pulses
+hw_timer_t * recharge_timer = NULL;
 
 // MUX (mutexes used for mutual exclusive access to isr variables)
 portMUX_TYPE mux_cap_full = portMUX_INITIALIZER_UNLOCKED;
 portMUX_TYPE mux_GMC_count = portMUX_INITIALIZER_UNLOCKED;
+portMUX_TYPE mux_hv = portMUX_INITIALIZER_UNLOCKED;
+
+void reschedule(hw_timer_t *timer, unsigned int cycles) {
+  timerAlarmWrite(timer, cycles, false);  // set alarm after <cycles>, do not repeat
+  timerWrite(timer, 0);
+  timerAlarmEnable(timer);
+}
+
+// How many HV capacitor charge pulses to generate for any later charge cycle.
+#define MAX_CHARGE_PULSES 333
+
+// How many HV capacitor charge pulses to generate for first charge.
+// After device power on, HV capacitor is completely empty.
+#define MAX_CHARGE_PULSES_INITIAL 3333
+
+void IRAM_ATTR isr_recharge() {
+  static int state = 0;
```

Comment:
i implemented a state machine here to keep the in-ISR time low.

the old code just used `delayMicroseconds()`, but we should not do that in a ISR.

Reply:
Could you please give a short comment in the code what the states represent? thx

Reply:
guess I'll use an enum to make it clearer.